File: 21553 - Sai Nandan D/Week 3/918_Maximum Sum Circular Subarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        max_sum, cur_max = nums[0], nums[0]
        min_sum, cur_min = nums[0], nums[0]
        total = nums[0]
        
        for n in nums[1:]:
            cur_max = max(n, cur_max+n)
            max_sum = max(max_sum, cur_max)
            cur_min = min(n, cur_min+n)
            min_sum = min(cur_min, min_sum)
            total += n
            
        return max(max_sum, total - min_sum) if max_sum > 0 else max_sum

File: 21561 - Suvam Tamang/Week 3/918_Maximum Sum Circular Subarray.java
class Solution {
    public int maxSubarraySumCircular(int[] nums) {
        int max_sum = nums[0], min_sum = nums[0], curr_sum1 = 0, curr_sum2 = 0, total = 0;
        int result;
        for(int n : nums){
            curr_sum1 += n;
            curr_sum2 += n;
            max_sum = Math.max(curr_sum1, max_sum);
            if(curr_sum1 < 0) curr_sum1 = 0;
            min_sum = Math.min(curr_sum2, min_sum);
            if(curr_sum2 > 0) curr_sum2 = 0;
            total += n;
        }
        if(max_sum < 0){
            result = max_sum;
        }
        else{
            result = Math.max(max_sum, total - min_sum);
        }
        return result;
    }
}
